pkg/service/agentservice.go

```diff
@@ -188,321 +166,207 @@ func (s *AgentHandler) HandleConnection(ctx context.Context, conn *websocket.Con
-			} else {
-				s.publisherRegistered = true
-				firstWorker = true
-			}
+	} else if w.JobType() == livekit.JobType_JT_ROOM {
+		numRooms++
+		if numRooms == 1 {
+			err = h.agentServer.RegisterJobRequestTopic(w.Namespace(), agent.RoomAgentTopic)
 		}
-	default:
-		s.mu.Unlock()
-		return errors.New("invalid job type")
 	}
-	s.mu.Unlock()
 
-	_, err := worker.sigConn.WriteServerMessage(&livekit.ServerMessage{
-		Message: &livekit.ServerMessage_Register{
-			Register: &livekit.RegisterWorkerResponse{
-				WorkerId:      worker.id,
-				ServerVersion: AgentServiceVersion,
-			},
-		},
-	})
 	if err != nil {
-		worker.logger.Errorw("failed to write server message", err)
+		w.Logger.Errorw("failed to register job request topic", err)
+		h.mu.Unlock()
+		w.Close() // Close the worker
+		return
 	}
 
-	if firstWorker && onRegistered != nil {
-		onRegistered(s)
+	h.namespaces[w.Namespace()] = &namespaceInfo{
+		numPublishers: numPublishers,
+		numRooms:      numRooms,
 	}
-	return nil
-}
 
-func (s *AgentHandler) handleAvailability(w *worker, msg *livekit.AvailabilityResponse) {
-	s.mu.Lock()
-	availabilityChan, ok := s.availability[msg.JobId]
-	s.mu.Unlock()
+	h.roomEnabled = h.roomAvailableLocked()
+	h.publisherEnabled = h.publisherAvailableLocked()
+	h.mu.Unlock()
 
-	if ok {
-		availabilityChan <- &availability{
-			workerID:  w.id,
-			available: msg.Available,
-		}
+	err = h.agentServer.PublishWorkerRegistered(context.Background(), "", &emptypb.Empty{})
+	if err != nil {
+		w.Logger.Errorw("failed to publish worker registered", err)
 	}
 }
 
-func (s *AgentHandler) handleJobUpdate(w *worker, msg *livekit.JobStatusUpdate) {
-	switch msg.Status {
-	case livekit.JobStatus_JS_SUCCESS:
-		w.logger.Debugw("job complete", "jobID", msg.JobId)
-	case livekit.JobStatus_JS_FAILED:
-		w.logger.Warnw("job failed", errors.New(msg.Error), "jobID", msg.JobId)
-	}
+func (h *AgentHandler) unregisterWorkerTopic(worker *agent.Worker) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
 
-	w.mu.Lock()
-	w.activeJobs--
-	w.mu.Unlock()
-}
+	info, ok := h.namespaces[worker.Namespace()]
+	if !ok {
+		return
+	}
 
-func (s *AgentHandler) handleStatus(w *worker, msg *livekit.UpdateWorkerStatus) {
-	s.mu.Lock()
-	defer s.mu.Unlock()
-
-	w.mu.Lock()
-	w.status = msg.Status
-	w.mu.Unlock()
-
-	switch w.jobType {
-	case livekit.JobType_JT_ROOM:
-		if s.roomRegistered && !s.roomAvailableLocked() {
-			s.roomRegistered = false
-			s.agentServer.DeregisterJobRequestTopic(s.roomTopic)
-		} else if !s.roomRegistered && s.roomAvailableLocked() {
-			if err := s.agentServer.RegisterJobRequestTopic(s.roomTopic); err != nil {
-				w.logger.Errorw("failed to register room agents", err)
-			} else {
-				s.roomRegistered = true
-			}
+	if worker.JobType() == livekit.JobType_JT_PUBLISHER {
+		info.numPublishers--
+		if info.numPublishers == 0 {
+			h.agentServer.DeregisterJobRequestTopic(worker.Namespace(), agent.PublisherAgentTopic)
```

Comment:
I guess the change in topics means all of the server instances would have to be updated at the same time. good to keep in mind.